tari-project/tari-dan
# refactor(consensus)!: block diffs, exec on propose and locks improvement

Description
---

- Generate and persist block diff for an accepted block
- generate a ProposedBlockChangeSet and persist for an accepted block
- State store read transactions no longer require a mutable reference
- Use PendingSubstateStore to track block changes and locks
- Remove previous locking code
- Allows previously conflicting local-only transactions to be sequenced using uncommitted outputs from other local only transactions as inputs (`Local-Only-Rules`)
- Fix multiple issues resulting in test flakiness
- Execute deferred transactions on propose and on prepare
- Adds deferred transaction test (currently for local-only case only)
- decide_what_to_vote is read-only
- Unignored some previously-flaky consensus tests
- Removed BlockTransactionExecutorBuilder trait
- Updated block sync

Motivation and Context
---

This PR improves locking, substate management and allows local-only transaction concurrency.

It adds the following concepts:
`BlockDiff` - contains the substate changes (UP or DOWN) for a block. Is used to commit state in a commit block 
`PendingSubstateStore` - tracks and validates changes to the substate store and decouples state from execution. It is used to generate a BlockDiff, track locks in the current block context, and to generate the state tree diff.
`ProposedBlockChangeSet` - contains all the pending changes to be committed when deciding to ACCEPT a new block
`TransactionExecution` - decouples execution results from a transaction record. These represent an execution within the context of a block. Once a transaction is finalized, the execution result is added to the final result in the transaction.

Possible issues:
Currently to fetch the correct transaction executions and locks for a block we need to limit the results to those that match the chain. This query will grow slower as the chain progresses. Bugs may still occur in the unhappy/malicious path.

How Has This Been Tested?
---
New deferred transaction consensus test, new unit tests, existing tests, removed ignore attribute from several consensus tests.

What process can a PR reviewer use to test or verify this change?
---
Submit transactions without specifying the version of the input. Submit local-only transactions concurrently that spend previous outputs from other local only transactions.

Breaking Changes
---


BREAKING CHANGE: database structure changed
